### PR TITLE
glTF Serializer: Added support for child node hierarchy

### PR DESCRIPTION
**Pull request by kcoley**
- Refactored to handle node hierarchies (parent and children nodes)

### CODE REVIEW

**Comment by kcoley** [comment]
Looking into the travis buld issue...

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 1]
```diff
@@ -161,7 +191,7 @@ module BABYLON.GLTF2 {
                     let scale = 1;
 
                     for (let j = 0; j < stride; ++j) {
-                        if (j === (stride - 1) && !useRightHandedSystem) {
+                        if (j === (stride - 1) && this.convertToRightHandedSystem) {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: This seems error-prone. Is it possible to calculate the min max after converting to right handed?

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 2]
```diff
@@ -186,121 +216,39 @@ module BABYLON.GLTF2 {
...
-
-                        if (useRightHandedSystem) {
-                            dataBuffer!.setFloat32(byteOff, meshAttributeArray![index + 2], true);
-                        }
-                        else {
-                            dataBuffer!.setFloat32(byteOff, -meshAttributeArray![index + 2], true);
-                        }
+            for (let k = 0; k < end; ++k) {
+                const index = k * strideSize;
+                dataBuffer!.setFloat32(byteOff, meshAttributeArray[index], true);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: ! isn't necessary

**Comment by kcoley** [comment]
Fixed the issue.  Should be good now.

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts (hunk 1 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFExporter.ts]: Yeah, I can do that

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 3]
```diff
@@ -186,121 +216,39 @@ module BABYLON.GLTF2 {
...
+                byteOff += 4;
 
-                        byteOff += 4;
-                    }
+                dataBuffer!.setFloat32(byteOff, meshAttributeArray[index + 1], true);
+                byteOff += 4;
+                if (strideSize > 2) {
+                    // Positions and Normals have 3 values.
+                    const scalez = (!this.convertToRightHandedSystem || vertexBufferKind === VertexBuffer.ColorKind) ? 1 : -1;
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: What happens with the fourth component of tangents?

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 4]
```diff
@@ -186,121 +216,39 @@ module BABYLON.GLTF2 {
...
+                dataBuffer.setFloat32(byteOff, meshAttributeArray[index], true);
+                byteOff += 4;
 
-                        dataBuffer!.setFloat32(byteOff, meshAttributeArray![index + 1], true);
-                        byteOff += 4;
+                dataBuffer.setFloat32(byteOff, meshAttributeArray[index + 1], true);
+                byteOff += 4;
+                if (strideSize > 2) {
+                    // Positions and Normals have 3 values.
+                    const scalez = (!this.convertToRightHandedSystem || vertexBufferKind === VertexBuffer.ColorKind) ? 1 : -1;
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: What happens with the fourth component of tangents?

Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFExporter.ts]: I think I'll refactor this part.  The fourth component of the tangent should be negated, but it is overlooked in this implementation.

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts (hunk 1 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFExporter.ts]: I now calculate the right-handed position vector before getting the min and max value from it.

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts (hunk 2 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFExporter.ts]: Thanks, I caught that in a previous revision.  There should no longer be any in the glTFExporter.ts file.

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts (hunk 4 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFExporter.ts]: I now use help functions to calculate the right handed vector for vector3 and vector4.

**Comment by kcoley** [comment]
@bghgary , I made the revisions you suggested and did a little more refactoring.

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 5]
```diff
@@ -143,41 +174,66 @@ module BABYLON.GLTF2 {
...
-                    let index = stride * i;
-                    let scale = 1;
+                    let indexOffset = positionStrideSize * i;
 
-                    for (let j = 0; j < stride; ++j) {
-                        if (j === (stride - 1) && !useRightHandedSystem) {
-                            scale = -1;
-                        }
-                        let num = scale * buff[index];
+                    const position = new Vector3(positions[indexOffset], positions[indexOffset + 1], positions[indexOffset + 2]);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: Vector3.FromArray

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 6]
```diff
@@ -186,121 +242,43 @@ module BABYLON.GLTF2 {
...
-                            dataBuffer!.setFloat32(byteOff, -meshAttributeArray![index + 2], true);
-                        }
-                        byteOff += 4;
-                        dataBuffer!.setFloat32(byteOff, meshAttributeArray![index + 3], true);
-                        byteOff += 4;
-                    }
-                    byteLength = meshAttributeArray.length * 4;
-                    break;
+                if (vertexBufferKind === VertexBuffer.PositionKind || vertexBufferKind === VertexBuffer.NormalKind) {
+                    const vertexData = new Vector3(meshAttributeArray[index], meshAttributeArray[index + 1], meshAttributeArray[index + 2]);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: Vector3.FromArray

File: serializers/src/glTF/2.0/babylon.glTFExporter.ts [hunk 7]
```diff
@@ -186,121 +242,43 @@ module BABYLON.GLTF2 {
...
-                        byteOff += 4;
-                        dataBuffer!.setFloat32(byteOff, meshAttributeArray![index + 2], true);
-                        byteOff += 4;
-                        dataBuffer!.setFloat32(byteOff, meshAttributeArray![index + 3], true);
-                        byteOff += 4;
-                    }
-                    byteLength = meshAttributeArray.length * 4;
-                    break;
+                else if (vertexBufferKind === VertexBuffer.TangentKind || vertexBufferKind === VertexBuffer.ColorKind) {
+                    const vertexData = new Vector4(meshAttributeArray[index], meshAttributeArray[index + 1], meshAttributeArray[index + 2], meshAttributeArray[index + 3]);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFExporter.ts]: ditto, Vector3.FromArray

**Pull request merged**